File: scratch/17.cpp
#include <bits/stdc++.h>
#define PB push_back
#define S string
#define VS vector<S>

using namespace std;

void p(VS s) { for (S c : s) { printf("%s\n", c); } }

VS letterCombonations(S digits) {
	if (digits == "") return {};
	VS mappings{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
	for (auto digit : digits) {
		VS extendCombonation;
		for (auto& cc : ans) {
			for (auto nc : mappings[digit - '2']) {
				extendCombonation.PB(cc + nc);
			}
		}
		swap(ans, extendCombonation);
	}
	return ans;
}

int main() {
	p(letterCombonations("123"));
}

File: scratch/17.py
from typing import List

def dfs(m, digits, path, ret):
    if not digits:
        ret.append(path)
        return
    for c in m[digits[0]]:
        dfs(m, digits[1:], path+c, ret)

def letterCombo(digits: str) -> List[str]:
    if not digits:
        return []
    ret = []
    m = {'2': 'abc', '3': 'def', '4': 'ghj', '5': 'klm', '6': 'nop', '7': 'qrs', '8': 'tuv', '9': 'wxyz'}
    dfs(m, digits, "", ret)
    return ret

print(letterCombo("123"))
